### PR TITLE
EG-4802 EG-4805 EG-4506 Bug fixes

### PR DESCRIPTION
**Pull request by MrMatthewLayton**
- Upgrading to requiring at least Corda 4.4 and built against 4.10
 - EG-4802 CreateEvolvableTokensFlowHandler updated to only allow signing when required by a transaction counter-party.
 -  EG-4805 + EG-4506 ObserverAwareFinalityFlow and ObserverAwareFinalityFlowHandler updated to allow extectedTxId, and checks that that received role complies with the transaction being sent.

### CODE REVIEW

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt [hunk 1]
```diff
@@ -1,25 +1,48 @@
...
-            TransactionRole.PARTICIPANT -> StatesToRecord.ONLY_RELEVANT
-            TransactionRole.OBSERVER -> StatesToRecord.ALL_VISIBLE
-        }
-        // If states are issued to self, then ReceiveFinalityFlow does not need to be invoked.
-        return if (!serviceHub.myInfo.isLegalIdentity(otherSession.counterparty)) {
-            subFlow(ReceiveFinalityFlow(otherSideSession = otherSession, statesToRecord = statesToRecord))
-        } else null
-    }
-}
+class ObserverAwareFinalityFlowHandler(val otherSession: FlowSession, val expectedTxId: SecureHash? = null) : FlowLogic<SignedTransaction?>() {
```
Comment (adelel1) [inline workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt]: I can't find anywhere in the code where the expectedTxId is set. So vulnerability is still there. The check below allows the txn if expectedTxId is null which it appears it always is.

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/evolvable/CreateEvolvableTokensFlowHandler.kt [hunk 2]
```diff
@@ -11,22 +11,26 @@ import net.corda.core.utilities.unwrap
...
+		val notification = otherSession.receive<CreateEvolvableTokensFlow.Notification>().unwrap { it }
 
-        // Resolve the creation transaction.
-        subFlow(ObserverAwareFinalityFlowHandler(otherSession))
-    }
+		// Sign the transaction proposal, if required
+		if (notification.signatureRequired) {
+			val signTransactionFlow = object : SignTransactionFlow(otherSession) {
+				override fun checkTransaction(stx: SignedTransaction) = requireThat {
+					require(stx.getMissingSigners().any { it in ourKeys }) {
```
Comment (adelel1) [inline workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/evolvable/CreateEvolvableTokensFlowHandler.kt]: Aren't more checks needed here? This is just checking the counterparty signature has been requested. But malicious node could craft any txn with the counterparty signature in it.

**Review by adelel1 — CHANGES_REQUESTED** [review]

File: Jenkinsfile [hunk 3]
```diff
@@ -46,17 +46,17 @@ pipeline {
...
-                timeout(60) {
-                    sh '''
-                        export ARTIFACTORY_USERNAME=\"\${ARTIFACTORY_CREDENTIALS_USR}\"
-                        export ARTIFACTORY_PASSWORD=\"\${ARTIFACTORY_CREDENTIALS_PSW}\"
-                        ./gradlew freighterTest --info --stacktrace --no-daemon
-                        '''
-                }
-            }
-        }
+//         stage('Freighter Tests') {
```
Comment (ronanbrowne) [inline Jenkinsfile]: just passing through as i've been asked to do some release support on this repo and was looking at this PR ...

These are commented out why ??

I know we have had issues with Freighter test stability in other places recently but are they a supported set of tests in this repo or not, If the answer is no should these just be removed as having them commented out just causes confusion for the next person that comes along? 

If they are support and should be working is there any plan to address this? If we are at a stage of accepting commenting them out to get PRs/ release to pass are they providing any value here?

Again i know relatively little about Freighter tests so maybe all this has been answered in conversations im not aware of!

Maybe @adelel1 has more context here .

**Review by ronanbrowne — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 3 above)
Comment (ronanbrowne) [reply Jenkinsfile]: My most recent commit on target branch adds a parameter to turn these on or off and are currently off by default due to stability issues, so you will pick that up when rebase and accept incoming changes

Comment (MrMatthewLayton) [reply Jenkinsfile]: @ronanbrowne I couldn't get them to work for love nor money.

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt (hunk 1 above)
Comment (adelel1) [reply workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt]: ObserverAwareFinalityFlowHandler is an inline flow. I have added the expected transaction id to those top level flow handlers which have the transaction id, these are the following...

CreateEvolvableTokensFlowHandler
UpdateEvolvableTokenFlowHandler
RedeemTokensFlowHandler

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/evolvable/CreateEvolvableTokensFlowHandler.kt (hunk 2 above)
Comment (adelel1) [reply workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/evolvable/CreateEvolvableTokensFlowHandler.kt]: The handler now only signs the transaction if it is one of the maintainers in one of the output states of the transaction.
This mirrors the client side which requests signatures from just the maintainers.

**Review by adelel1 — APPROVED** [review]

File: Jenkinsfile (hunk 3 above)
Comment (adelel1) [reply Jenkinsfile]: Thanks, I don't really have any more context than you. I agree, I would leave them in uncommented and switched off by default. Until they are eventually enabled. Not aware of them not being supported. No current plans to investigate them though.

**Comment by adelel1** [comment]
Hi @MrMatthewLayton I answered my own comments and updated the code. If you get a chance would you be able to do a quick review of the updates I did. Many Thanks,
(I couldn't add you as a reviewer presumably cause you are the author of PR)

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/evolvable/CreateEvolvableTokensFlowHandler.kt (hunk 2 above)
Comment (rick-r3) [reply workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/evolvable/CreateEvolvableTokensFlowHandler.kt]: This line of code appears to be redundant.   It seems to be checked in `SignTransactionFlow` anyway.  i.e. that the the missing signers has one of our keys in it and that the key we are asked to sign with is in it.

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/evolvable/CreateEvolvableTokensFlowHandler.kt [hunk 4]
```diff
@@ -11,22 +13,34 @@ import net.corda.core.utilities.unwrap
 
 /** In-line counter-flow to [CreateEvolvableTokensFlow]. */
 class CreateEvolvableTokensFlowHandler(val otherSession: FlowSession) : FlowLogic<Unit>() {
-    @Suspendable
-    override fun call() {
-        // Receive the notification
-        val notification = otherSession.receive<CreateEvolvableTokensFlow.Notification>().unwrap { it }
+	@Suspendable
+	override fun call() {
+		val ourKeys = serviceHub.keyManagementService.filterMyKeys(serviceHub.keyManagementService.keys)
```
Comment (rick-r3) [inline workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/evolvable/CreateEvolvableTokensFlowHandler.kt]: Not needed if line 28 not needed (which I don't think it is).

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt [hunk 5]
```diff
@@ -1,25 +1,48 @@
...
+	@Suspendable
+	override fun call(): SignedTransaction? {
+
+		val ourKeys = serviceHub.keyManagementService.filterMyKeys(serviceHub.keyManagementService.keys)
+
+		val role = otherSession.receive<TransactionRole>().unwrap { it }
+		val statesToRecord = role.toStatesToRecord()
+
+		return if (otherSession.counterparty.owningKey in ourKeys) null else {
```
Comment (rick-r3) [inline workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt]: I think the original logic that used `serviceHub.myInfo.isLegalIdentity(otherSession.counterparty)` is better and more efficient.

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt [hunk 6]
```diff
@@ -1,25 +1,48 @@
...
-        return if (!serviceHub.myInfo.isLegalIdentity(otherSession.counterparty)) {
-            subFlow(ReceiveFinalityFlow(otherSideSession = otherSession, statesToRecord = statesToRecord))
-        } else null
-    }
-}
+class ObserverAwareFinalityFlowHandler(val otherSession: FlowSession, val expectedTxId: SecureHash? = null) : FlowLogic<SignedTransaction?>() {
+	@Suspendable
+	override fun call(): SignedTransaction? {
+
+		val ourKeys = serviceHub.keyManagementService.filterMyKeys(serviceHub.keyManagementService.keys)
```
Comment (rick-r3) [inline workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt]: Don't do this.  It loads every identity the node knows about , and then looks them up in the database.  Thankfully don't think it is needed if change line 23 as per below comment.

**Review by MrMatthewLayton — COMMENTED** [review]
LGTM

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/evolvable/CreateEvolvableTokensFlowHandler.kt (hunk 2 above)
Comment (adelel1) [reply workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/flows/evolvable/CreateEvolvableTokensFlowHandler.kt]: Updated - removed redundant code. 
Yes the SignTransactionFlow receives the signing key expected to be used. 
Also redundant code was using getMissingSigners, where as code in SignTransactionFlow checks against requiredSigningKeys. (Processing could have been different if using composite keys).

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt (hunk 5 above)
Comment (adelel1) [reply workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt]: Reverted.

File: workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt (hunk 6 above)
Comment (adelel1) [reply workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/flows/finality/ObserverAwareFinalityFlowHandler.kt]: Yes, removed now.

**Comment by adelel1** [comment]
To maintain Java client compatibility added a @JvmOverloads to the ObserverAwareFinalityFlowHandler constructor.

**Review by rick-r3 — APPROVED** [review]
LGTM

File: Jenkinsfile (hunk 3 above)
Comment (MrMatthewLayton) [reply Jenkinsfile]: I think we need @roastario back as he's likely to know the most about the freighter tests.

**Review by rick-r3 — COMMENTED** [review]
I think you could make an argument for raising the platform versions of the cordapps even higher, since we have just increased from 4.3 to 4.4 here and 4.5 is the last one we will security fix by the time this ships, but I don't have a strong opinion, so okay to leave as is.

**Comment by rick-r3** [comment]
Update the description to reflect 4.10 etc.

**Review by ronanbrowne — APPROVED** [review]
lgtm ( build changes)

**Review by rick-r3 — APPROVED** [review]

**Pull request merged**